File: overlay/widgets/flags.py
"""
Flags -- a big, standalone flag banner.

Draws the current race flag (caution, black, blue, white, checkered, ...) as a
single bold banner inside the same dark rounded panel the other widgets use, with
the diagonal-slash (or checkerboard) texture and label plate borrowed from the
dash flag so it sits naturally alongside the rest of the overlay. When nothing is
flying it shows a calm muted state.

Expected data dict:
    flag          flag name from the app's session-flag logic, or None
    flag_context  optional detail line (e.g. "1 lap to green")
"""

from __future__ import annotations

from PyQt6.QtCore import QPointF, QRectF, Qt
from PyQt6.QtGui import (QColor, QFontMetricsF, QPainter,
                         QPainterPath, QPen)
from PyQt6.QtWidgets import QSizePolicy, QWidget

from .. import config
from .chrome import col, draw_card, draw_dark_cell
from .fonts import tfont

_SECTION = "flags"

# flag name -> (label, background color key, text color key).
_SPEC = {
    "yellow": ("CAUTION", "flag_yellow", "flag_yellow_text"),
    "black": ("BLACK FLAG", "flag_black", "flag_black_text"),
    "meatball": ("MEATBALL", "flag_meatball", "flag_meatball_text"),
    "furled": ("WARNING", "flag_furled", "flag_furled_text"),
    "dq": ("DISQUALIFIED", "flag_dq", "flag_dq_text"),
    "green": ("GREEN", "flag_green", "flag_green_text"),
    "white": ("LAST LAP", "flag_white_bg", "flag_white_text"),
    "red": ("RED FLAG", "flag_red", "flag_red_text"),
    "blue": ("LET BY", "flag_blue", "flag_blue_text"),
    "debris": ("DEBRIS", "flag_debris", "flag_debris_text"),
    "crossed": ("HALFWAY", "flag_crossed", "flag_crossed_text"),
    "checkered": ("FINISH", "flag_checker_bg", "flag_checker_text"),
}


class FlagsWidget(QWidget):
    def __init__(self, parent=None):
        super().__init__(parent)
        self.data: dict = {}
        self.setMinimumSize(180, 80)
        self.setSizePolicy(QSizePolicy.Policy.Expanding, QSizePolicy.Policy.Expanding)

    def set_data(self, data: dict) -> None:
        data = data or {}
        if data == self.data:
            return
        self.data = data
        self.update()

    def _cfg(self) -> dict:
        return config.CFG["flags"]

    def _col(self, key: str) -> QColor:
        return col(key, _SECTION)

    def paintEvent(self, event) -> None:  # noqa: N802
        flag = (self.data or {}).get("flag")
        incident_warn = (self.data or {}).get("incident_warn")
        if flag is None and not (self.data or {}).get("edit") and not incident_warn:
            return
        config.use_section(_SECTION)
        p = QPainter(self)
        p.setRenderHint(QPainter.RenderHint.Antialiasing)
        w, h = float(self.width()), float(self.height())
        draw_card(p, w, h, _SECTION)

        pad = max(6.0, min(w, h) * 0.12)
        if flag is None and incident_warn:
            rect = QRectF(pad, pad, w - 2 * pad, h - 2 * pad)
            draw_dark_cell(p, rect, _SECTION, radius=min(rect.height() * 0.34, 22.0))
            p.setFont(tfont(rect.height() * 0.24, True))
            p.setPen(self._col("flag_furled"))
            sec = str((self.data or {}).get("secondary") or "")
            p.drawText(rect, Qt.AlignmentFlag.AlignCenter, sec or "Incident warning")
            return
        self._draw_banner(p, QRectF(pad, pad, w - 2 * pad, h - 2 * pad), flag,
                          (self.data or {}).get("flag_context"),
                          (self.data or {}).get("secondary"))

    def _draw_banner(self, p, rect: QRectF, flag, context=None,
                     secondary=None) -> None:
        spec = _SPEC.get(flag)
        r = min(rect.height() * 0.34, 22.0)
        if spec is None:
            draw_dark_cell(p, rect, _SECTION, radius=r)
            p.setFont(tfont(rect.height() * 0.26))
            p.setPen(self._col("idle_text"))
            p.drawText(rect, Qt.AlignmentFlag.AlignCenter,
                       str(self._cfg().get("idle_text", "TRACK CLEAR")))
            return

        label, bg_key, fg_key = spec
        bg = self._col(bg_key)
        fg = self._col(fg_key)
        p.setBrush(bg)
        p.setPen(QPen(QColor(255, 255, 255, 45), 1))
        p.drawRoundedRect(rect, r, r)

        # Texture, clipped to the banner: checkerboard for the finish flag,
        # diagonal slashes otherwise (matches the dash flag treatment).
        p.save()
        clip = QPainterPath()
        clip.addRoundedRect(rect, r, r)
        p.setClipPath(clip)
        if flag == "checkered":
            self._draw_checker(p, rect, fg)
        else:
            hatch = QColor(fg)
            hatch.setAlpha(64)
            p.setPen(QPen(hatch, max(2.0, rect.height() * 0.10)))
            step = rect.height() * 0.5
            x = rect.left() - rect.height()
            while x < rect.right() + rect.height():
                p.drawLine(QPointF(x, rect.bottom()),
                           QPointF(x + rect.height(), rect.top()))
                x += step
        p.restore()

        # Rounded plate behind the label so the texture frames it.
        context = str(context).strip() if context else ""
        if context:
            title_font = tfont(rect.height() * 0.22, True)
            sub_font = tfont(rect.height() * 0.15, False)
            avail = rect.width() * 0.82
            tw = max(QFontMetricsF(title_font).horizontalAdvance(label),
                     QFontMetricsF(sub_font).horizontalAdvance(context))
            if tw > avail and tw > 0:
                scale = avail / tw
                title_font = tfont(rect.height() * 0.22 * scale, True)
                sub_font = tfont(rect.height() * 0.15 * scale, False)
                tw = max(QFontMetricsF(title_font).horizontalAdvance(label),
                         QFontMetricsF(sub_font).horizontalAdvance(context))
            plate_pad = rect.height() * 0.24
            plate_h = min(rect.height() * 0.72,
                            title_font.pixelSize() * 2.6 + sub_font.pixelSize() * 1.2)
            plate = QRectF(rect.center().x() - tw / 2 - plate_pad,
                           rect.center().y() - plate_h / 2,
                           tw + plate_pad * 2, plate_h)
            p.setPen(Qt.PenStyle.NoPen)
            p.setBrush(bg)
            p.drawRoundedRect(plate, plate.height() / 2, plate.height() / 2)
            title_y = rect.center().y() - plate_h * 0.16
            sub_y = rect.center().y() + plate_h * 0.18
            p.setFont(title_font)
            p.setPen(fg)
            p.drawText(QRectF(rect.left(), title_y - title_font.pixelSize() * 0.5,
                              rect.width(), title_font.pixelSize() * 1.4),
                       Qt.AlignmentFlag.AlignCenter, label)
            sub_fg = QColor(fg)
            sub_fg.setAlpha(min(255, int(fg.alpha() * 0.88)))
            p.setFont(sub_font)
            p.setPen(sub_fg)
            p.drawText(QRectF(rect.left(), sub_y - sub_font.pixelSize() * 0.5,
                              rect.width(), sub_font.pixelSize() * 1.4),
                       Qt.AlignmentFlag.AlignCenter, context)
            return

        font = tfont(rect.height() * 0.30)
        avail = rect.width() * 0.82
        tw = QFontMetricsF(font).horizontalAdvance(label)
        if tw > avail and tw > 0:
            font = tfont(rect.height() * 0.30 * avail / tw)
            tw = QFontMetricsF(font).horizontalAdvance(label)
        plate_pad = rect.height() * 0.28
        plate_h = min(rect.height() * 0.62, font.pixelSize() * 1.9)
        plate = QRectF(rect.center().x() - tw / 2 - plate_pad,
                       rect.center().y() - plate_h / 2,
                       tw + plate_pad * 2, plate_h)
        p.setPen(Qt.PenStyle.NoPen)
        p.setBrush(bg)
        p.drawRoundedRect(plate, plate.height() / 2, plate.height() / 2)
        p.setFont(font)
        p.setPen(fg)
        p.drawText(rect, Qt.AlignmentFlag.AlignCenter, label)

    @staticmethod
    def _draw_checker(p, rect: QRectF, color) -> None:
        rows = 3
        sq = rect.height() / rows
        cols = int(rect.width() / sq) + 2
        cell = QColor(color)
        cell.setAlpha(190)
        p.setPen(Qt.PenStyle.NoPen)
        p.setBrush(cell)
        for ri in range(rows):
            for ci in range(cols):
                if (ri + ci) % 2 == 0:
                    p.drawRect(QRectF(rect.left() + ci * sq,
                                      rect.top() + ri * sq, sq, sq))

File: overlay-rs/crates/gridglance-overlay/src/widgets/flags.rs
//! Flags banner — plate + hatch/checker (Python `flags.py`).

use super::WidgetCtx;
use crate::chrome::{color_with_alpha, draw_card, draw_dark_cell, full_rect, label};
use egui::{Align2, Color32, CornerRadius, Pos2, Rect, Stroke, StrokeKind, Ui, Vec2};

const SECTION: &str = "flags";

const SPEC: &[(&str, &str, &str, &str)] = &[
    ("yellow", "CAUTION", "flag_yellow", "flag_yellow_text"),
    ("black", "BLACK FLAG", "flag_black", "flag_black_text"),
    ("meatball", "MEATBALL", "flag_meatball", "flag_meatball_text"),
    ("furled", "WARNING", "flag_furled", "flag_furled_text"),
    ("dq", "DISQUALIFIED", "flag_dq", "flag_dq_text"),
    ("green", "GREEN", "flag_green", "flag_green_text"),
    ("white", "LAST LAP", "flag_white_bg", "flag_white_text"),
    ("red", "RED FLAG", "flag_red", "flag_red_text"),
    ("blue", "LET BY", "flag_blue", "flag_blue_text"),
    ("debris", "DEBRIS", "flag_debris", "flag_debris_text"),
    ("crossed", "HALFWAY", "flag_crossed", "flag_crossed_text"),
    ("checkered", "FINISH", "flag_checker_bg", "flag_checker_text"),
];

fn text_width(ui: &Ui, text: &str, size: f32, bold: bool) -> f32 {
    let font = egui::FontId::proportional(size.max(1.0));
    let galley = ui.fonts(|f| f.layout_no_wrap(text.to_owned(), font, Color32::WHITE));
    let w = galley.size().x;
    // Fake-bold widens slightly.
    if bold {
        w + 0.9
    } else {
        w
    }
}

fn draw_checker(painter: &egui::Painter, rect: Rect, color: Color32) {
    let rows = 3.0_f32;
    let sq = rect.height() / rows;
    let cols = (rect.width() / sq).ceil() as i32 + 2;
    let cell = color_with_alpha(color, 190);
    for ri in 0..3 {
        for ci in 0..cols {
            if (ri + ci) % 2 == 0 {
                let x = rect.left() + ci as f32 * sq;
                let y = rect.top() + ri as f32 * sq;
                painter.rect_filled(
                    Rect::from_min_size(
                        Pos2::new(x, y),
                        Vec2::new(sq.min(rect.right() - x), sq.min(rect.bottom() - y)),
                    ),
                    CornerRadius::ZERO,
                    cell,
                );
            }
        }
    }
}

fn draw_hatch(painter: &egui::Painter, rect: Rect, fg: Color32) {
    let hatch = color_with_alpha(fg, 64);
    let step = rect.height() * 0.5;
    let pen = Stroke::new((rect.height() * 0.10).max(2.0), hatch);
    let mut x = rect.left() - rect.height();
    while x < rect.right() + rect.height() {
        painter.line_segment(
            [
                Pos2::new(x, rect.bottom()),
                Pos2::new(x + rect.height(), rect.top()),
            ],
            pen,
        );
        x += step;
    }
}

pub fn paint(ui: &mut Ui, ctx: &mut WidgetCtx<'_>) {
    let f = ctx.frame;
    let have_secondary = f
        .secondary
        .as_deref()
        .map(|s| !s.is_empty())
        .unwrap_or(false);
    if f.flag.is_none() && !ctx.edit_mode && !f.incident_warn && !have_secondary {
        let _ = full_rect(ui);
        return;
    }
    let rect = full_rect(ui);
    draw_card(ui, ctx.cfg, SECTION, rect);
    let pad = (rect.height() * 0.12).max(6.0);
    let inner = rect.shrink(pad);

    if f.flag.is_none() && f.incident_warn {
        draw_dark_cell(
            ui,
            ctx.cfg,
            SECTION,
            inner,
            (inner.height() * 0.34).min(22.0),
        );
        let msg = f.secondary.as_deref().unwrap_or("Incident warning");
        label(
            ui,
            inner.center(),
            Align2::CENTER_CENTER,
            msg,
            inner.height() * 0.24,
            ctx.cfg.color(SECTION, "flag_furled", "#caa23a"),
            true,
        );
        return;
    }

    let flag = f.flag.as_deref().unwrap_or("");
    if let Some((_, title, bgk, fgk)) = SPEC.iter().find(|(k, ..)| *k == flag) {
        let bg = ctx.cfg.color(SECTION, bgk, "#46df7a");
        let fg = ctx.cfg.color(SECTION, fgk, "#141414");
        let r = (inner.height() * 0.34).min(22.0);
        ui.painter()
            .rect_filled(inner, CornerRadius::same(r as u8), bg);
        ui.painter().rect_stroke(
            inner,
            CornerRadius::same(r as u8),
            Stroke::new(1.0_f32, Color32::from_rgba_unmultiplied(255, 255, 255, 45)),
            StrokeKind::Inside,
        );

        // Texture clipped to banner.
        {
            let painter = ui.painter().with_clip_rect(inner);
            if flag == "checkered" {
                draw_checker(&painter, inner, fg);
            } else {
                draw_hatch(&painter, inner, fg);
            }
        }

        let context = f
            .flag_context
            .as_deref()
            .or(f.secondary.as_deref())
            .map(str::trim)
            .filter(|s| !s.is_empty())
            .unwrap_or("");

        if !context.is_empty() {
            let mut title_sz = inner.height() * 0.22;
            let mut sub_sz = inner.height() * 0.15;
            let avail = inner.width() * 0.82;
            let mut tw = text_width(ui, title, title_sz, true)
                .max(text_width(ui, context, sub_sz, false));
            if tw > avail && tw > 0.0 {
                let scale = avail / tw;
                title_sz *= scale;
                sub_sz *= scale;
                tw = text_width(ui, title, title_sz, true)
                    .max(text_width(ui, context, sub_sz, false));
            }
            let plate_pad = inner.height() * 0.24;
            let plate_h = (inner.height() * 0.72).min(title_sz * 2.6 + sub_sz * 1.2);
            let plate = Rect::from_center_size(
                inner.center(),
                Vec2::new(tw + plate_pad * 2.0, plate_h),
            );
            ui.painter().rect_filled(
                plate,
                CornerRadius::same((plate_h * 0.5).round().clamp(0.0, 255.0) as u8),
                bg,
            );
            label(
                ui,
                Pos2::new(inner.center().x, inner.center().y - plate_h * 0.16),
                Align2::CENTER_CENTER,
                title,
                title_sz,
                fg,
                true,
            );
            label(
                ui,
                Pos2::new(inner.center().x, inner.center().y + plate_h * 0.18),
                Align2::CENTER_CENTER,
                context,
                sub_sz,
                color_with_alpha(fg, ((fg.a() as f32) * 0.88) as u8),
                false,
            );
        } else {
            let mut font_sz = inner.height() * 0.30;
            let avail = inner.width() * 0.82;
            let mut tw = text_width(ui, title, font_sz, true);
            if tw > avail && tw > 0.0 {
                font_sz *= avail / tw;
                tw = text_width(ui, title, font_sz, true);
            }
            let plate_pad = inner.height() * 0.28;
            let plate_h = (inner.height() * 0.62).min(font_sz * 1.9);
            let plate = Rect::from_center_size(
                inner.center(),
                Vec2::new(tw + plate_pad * 2.0, plate_h),
            );
            ui.painter().rect_filled(
                plate,
                CornerRadius::same((plate_h * 0.5).round().clamp(0.0, 255.0) as u8),
                bg,
            );
            label(
                ui,
                inner.center(),
                Align2::CENTER_CENTER,
                title,
                font_sz,
                fg,
                true,
            );
        }
    } else {
        // Idle plate (Python: dark cell + idle_text).
        let r = (inner.height() * 0.34).min(22.0);
        draw_dark_cell(ui, ctx.cfg, SECTION, inner, r);
        let idle = if let Some(sec) = f.secondary.as_deref().filter(|s| !s.is_empty()) {
            sec.to_string()
        } else {
            // Config string key shares the name `idle_text` with the color key
            // under `colors.idle_text` — str_key reads the section field.
            ctx.cfg.str_key(SECTION, "idle_text", "TRACK CLEAR")
        };
        label(
            ui,
            inner.center(),
            Align2::CENTER_CENTER,
            &idle,
            inner.height() * 0.26,
            ctx.cfg.color(SECTION, "idle_text", "#9fb0a4"),
            false,
        );
    }
}
